File: 792-BinarySearch/792-BinarySearch.py
# Last updated: 11/2/2025, 8:22:22 PM
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        #list is sorted in asecnding order, so we can use binary search
        left = 0 
        right = len(nums) - 1 

        #edge case: <=, if array only has one element we still want to check if the target exists 
        while left <= right:
            mid = left + (right-left)//2 

            if nums[mid] == target:
                return mid #return idx of where the target is 
            elif nums[mid] > target: #that means the target cannot be at mid or the right 
                right = mid - 1 
            else: #target cannot be at mid or the left
                left = mid + 1 
        
        return -1 #target does not exist in the list 

        #Time complexity: O(logn)
        #Space complexity: O(1)

File: 792-BinarySearch/792-BinarySearch.java
// Last updated: 11/2/2025, 8:22:25 PM
class Solution {
    public int search(int[] nums, int target) {
        
        int left = 0;
        int right = nums.length-1;

        while(left<=right){

            int mid = (left+right)/2;
            if(nums[mid] == target){
                return mid; 
            }
            else if(nums[mid]>target){
                right = mid-1;
            }
            else if(nums[mid]<target){
                left = mid+1;
            }
        
        }
        return -1;

    }
}

//Time complexity: O(log⁡n). nums is divided into half each time. In the worst-case scenario, we need to cut nums until the range has no element, and it takes logarithmic time to reach this break condition.
//Space complexity: O(1). During the loop, we only need to record three indexes, left, right, and mid, they take constant space.